Issue #6 Prevent duplicate entries and implement form validation

katia-utochkina opened the pull request:
## Description

This PR prevents the user from adding duplicate entries to the shopping list. It compares db items with the ones submitted by the user via input and makes sure that items with the same name do not get added to the shopping list. If the user tries to submit a duplicate item, they are shown an error message

## Related Issue

<!-- If you write "closes" followed by the Github issue number, it will automatically close the issue for you when the PR merges -->

## Acceptance Criteria

- [x] Show an error message if the user tries to submit a new item that has the exact same name as an existing item
- [x] Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalisation has been normalised and punctuation has been removed
- [x] The user’s original input should be what gets saved in the database

## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Testing Steps / QA Criteria

* From your terminal, pull down this branch with `git pull origin ku-ci-prevent-duplicates` and check that branch out with `git checkout ku-ci-prevent-duplicates` 
* Then `npm start` to launch the app
* Clear local storage. Generate a new token for setting up a shopping list. Submit “Add " form with no input to make sure an empty entry doesn't get added to the list
* Enter an item that is already on the list to make sure you're alerted that the item is already on the list. Use capitalisation and punctuation (f.ex  `To?m/a&t@oe$s%`). 
* Submit an item with no frequency selected to make sure you're alerted that this field is required. 
* Make sure duplicate items do not appear in list view and in the db

C-Zekeri commented on src/App.js:
```diff
@@ -1,35 +1,63 @@
   };
+
+  const [shoppingList, loading, error] = useCollectionData(
+    db.collection('lists').where('token', '==', token),
```
Not sure I understand what this bit of code is doing, but shouldn't the collection be `shoppinglist` instead of `lists`?

C-Zekeri commented on src/components/AddView.js:
```diff
@@ -1,91 +1,140 @@
+  const addListItem = async (item, collectionId) => {
+    let { name, howSoon, lastPurchasedDate } = item;
+    try {
+      db.collection('lists')
```
Again, I'm not sure why we are adding items to the list collection instead of the shoppinglist collection, as was originally done.

katia-utochkina replied:
we've renamed ``shoppinglist`` collection into ``lists``  because in the app, we will have a bunch of users creating a bunch of lists.

C-Zekeri commented on src/App.js:
```diff
@@ -1,35 +1,63 @@
   };
+
+  const [shoppingList, loading, error] = useCollectionData(
+    db.collection('lists').where('token', '==', token),
```
Not sure I understand what this bit of code is doing, but shouldn't the collection be `shoppinglist` instead of `lists`?

katia-utochkina replied:
Hi @C-Zekeri , you're right that's how it was named before. But it would probably be more logical to have  ``shoppinglist`` changed to ``lists`` - because in the app, we will have a bunch of users creating a bunch of lists.

C-Zekeri replied:
If it's about naming, you could just edit the name of the collection in the database, instead of writing to a new collection, yeah?

C-Zekeri reviewed (commented):
Hi! This is a great implementation, well-done!
Just have a question about the change in the collection the app is writing to. Not sure if I misunderstood, so could you please explain to me?

faysvas commented on src/components/AddView.js:
```diff
@@ -1,91 +1,137 @@
+        return;
+      }
+
+      if (item.name === '' && item.howSoon === '') {
+        setIsNameFieldEmpty(true);
+        setIsFrequencyFieldEmpty(true);
+        return;
+      }
```
Could this be omitted since since you check if `name` and `howSoon` are empty in the next lines?

katia-utochkina replied:
Hi @faysvas . We can't omit it. Without the check, when both fields are empty, an error message will be displayed for the first field only

faysvas commented on src/components/ListView.js:
```diff
@@ -1,28 +1,29 @@
+        {shoppingList &&
+          shoppingList.map((list) =>
+            list.items
+              .map((item) => item)
```
Why is this extra map added? It just seems to go through all the items and return them?

katia-utochkina replied:
Good catch, I've missed that

faysvas commented on src/lib/utils.js:
```diff
@@ -0,0 +1,5 @@
+const transformUserInput = (item) => {
+  return item.replace(/[^a-zA-Z ]/g, '').replace(/\s\s+/g, ' ');
+};
```
This is a very nice touch, but what if I want to buy "M&M's" or "100% natural juice" :slightly_smiling_face: ?
Also, this kind of arrow functions (they can be called one-liners) can also be written like this:
```
const transformUserInput = item => item.replace(/[^a-zA-Z ]/g, '').replace(/\s\s+/g, ' ')
```
They are both the exact same thing, so you can use whatever is more readable for you :slightly_smiling_face:

katia-utochkina replied:
We thought removing symbols like ``'%&`` was a requirement, too ? If you enter ``M&M's`` or `` 100% natural Juice``, it'll be transformed into ``MMs`` and ``Natural Juice``
If not this way, how would you suggest implementing this?

faysvas commented on src/App.css:
```diff
@@ -20,92 +20,196 @@ h2,
+  font-family: inherit;
+  font-weight: 400;
+  font-style: italic;
+  color: inherit;
```
You can skip inheriting color, it's a property that's inherited by default. Inheritance in CSS is very interesting https://developer.mozilla.org/en-US/docs/Web/CSS/inheritance

katia-utochkina replied:
We removed it :slightly_smiling_face:

faysvas reviewed (commented):
I know this is still WIP but I didn't want to wait until the last minute to add my comments:slightly_smiling_face: . The code looks great and the CSS is very stylish :smiley:
 Just one note: Spaces aren't trimmed when I enter items. So it allows me to enter "Tomato   " even if I already have "Tomato" entered.

katia-utochkina commented:
> I know this is still WIP but I didn't want to wait until the last minute to add my comments🙂 . The code looks great and the CSS is very stylish 😃
> Just one note: Spaces aren't trimmed when I enter items. So it allows me to enter "Tomato " even if I already have "Tomato" entered.

We had empty space removal before but I deleted it by accident amid refactoring🙂 .. Just fixed it

faysvas commented on src/lib/utils.js:
```diff
@@ -0,0 +1,8 @@
+    .toLowerCase()
+    .trim();
+
+export default transformUserInput;
```
Good job on adding the `trim` and `toLowerCase` :+1: . For the regex, you can also[ use `/[\W_]/g` which finds all non word characters and underscores ](https://stackoverflow.com/questions/35324229/how-to-interpret-this-regular-expression-w-g). 
The issue here is that you the third requirement of the issue isn't satisfied:
**"The user’s original input should be what gets saved in the database".**
So, for example, if I want to save an item named "potato!", I should be allowed to save it exactly as it is, no modification should be done to the input (well, except for trimming maybe). But if I try to add "Potato" I shouldn't be able to add it :slightly_smiling_face: 
To do this, you should only compare the user's input to the existing items, without actually modifying the input.
